### PR TITLE
New reindex spring batch operation

### PR DESCRIPTION
**Pull request by fil512**

### CODE REVIEW

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/DeleteExpungeProvider.java [hunk 1]
```diff
@@ -52,18 +43,7 @@ public IBaseParameters deleteExpunge(
...
-			}
-			JobExecution jobExecution = myDeleteExpungeJobSubmitter.submitJob(batchSize, theRequestDetails, urls);
-			IBaseParameters retval = ParametersUtil.newInstance(myFhirContext);
-			ParametersUtil.addParameterToParametersLong(myFhirContext, retval, ProviderConstants.OPERATION_DELETE_EXPUNGE_RESPONSE_JOB_ID, jobExecution.getJobId());
-			return retval;
-		} catch (JobParametersInvalidException e) {
-			throw new InvalidRequestException("Invalid job parameters: " + e.getMessage(), e);
-		}
+		List<String> urls = theUrlsToDeleteExpunge.stream().map(IPrimitiveType::getValue).collect(Collectors.toList());
+		return super.processUrls(urls, getBatchSize(theBatchSize), theRequestDetails);
```
Comment (michaelabuckley) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/DeleteExpungeProvider.java:47]: OH. Here's the yoyo-call.  I got lost.  Can we do this by composition instead?

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ProviderConstants.java [hunk 2]
```diff
@@ -147,4 +147,29 @@
...
+
+	/**
+	 * Operation name for the $delete-expunge operation
+	 */
+	public static final String OPERATION_REINDEX = "$reindex";
+
+	/**
+	 * url of resources to delete for the $delete-expunge operation
+	 */
+	public static final String OPERATION_REINDEX_PARAM_URL = "url";
```
Comment (michaelabuckley) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ProviderConstants.java:159]: Can we share these params - url and batchSize can be shared with delete-expunge.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ProviderConstants.java [hunk 3]
```diff
@@ -147,4 +147,29 @@
...
+
+	/**
+	 * Whether all resource types should be reindexed
+	 */
+	public static final String OPERATION_REINDEX_PARAM_EVERYTHING = "everything";
+
+	/**
+	 * The Spring Batch job id of the delete expunge job created by a $delete-expunge operation
+	 */
+	public static final String OPERATION_REINDEX_RESPONSE_JOB_ID = "jobId";
```
Comment (michaelabuckley) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ProviderConstants.java:174]: Is this delete, reindex, or generic?

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ReindexProvider.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		@OperationParam(name = ProviderConstants.OPERATION_REINDEX_PARAM_BATCH_SIZE, typeName = "decimal", min = 0, max = 1) IPrimitiveType<BigDecimal> theBatchSize,
+		@OperationParam(name = ProviderConstants.OPERATION_REINDEX_PARAM_EVERYTHING, typeName = "boolean", min = 0, max = 1) IPrimitiveType<Boolean> theEverything,
+		RequestDetails theRequestDetails
+	) {
+		Boolean everything = theEverything != null && theEverything.getValue();
+		@Nullable Integer batchSize = getBatchSize(theBatchSize);
+		if (everything) {
+			return processEverything(batchSize, theRequestDetails);
+		} else {
+			List<String> urls = theUrlsToReindex.stream().map(IPrimitiveType::getValue).collect(Collectors.toList());
```
Comment (michaelabuckley) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ReindexProvider.java:60]: Can theUrlsToReindex be null?

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ReindexProvider.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	@Operation(name = ProviderConstants.OPERATION_REINDEX, idempotent = false)
+	public IBaseParameters Reindex(
+		@OperationParam(name = ProviderConstants.OPERATION_REINDEX_PARAM_URL, typeName = "string", min = 0, max = 1) List<IPrimitiveType<String>> theUrlsToReindex,
+		@OperationParam(name = ProviderConstants.OPERATION_REINDEX_PARAM_BATCH_SIZE, typeName = "decimal", min = 0, max = 1) IPrimitiveType<BigDecimal> theBatchSize,
+		@OperationParam(name = ProviderConstants.OPERATION_REINDEX_PARAM_EVERYTHING, typeName = "boolean", min = 0, max = 1) IPrimitiveType<Boolean> theEverything,
+		RequestDetails theRequestDetails
+	) {
+		Boolean everything = theEverything != null && theEverything.getValue();
+		@Nullable Integer batchSize = getBatchSize(theBatchSize);
+		if (everything) {
```
Comment (michaelabuckley) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ReindexProvider.java:57]: Do we care enough to check if they set everything=true _AND_ passed some urls?

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_5_0/2845-reindex-operation.yaml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+---
+type: add
+issue: 2845
+title: "Added new $reindex operation with similar syntax to $delete-expunge that creates a spring-batch job to reindex selected resources."
```
Comment (michaelabuckley) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_5_0/2845-reindex-operation.yaml:4]: This is a bit light for the size of the change.  
```suggestion
title: "Added new `$reindex` operation with similar syntax to `$delete-expunge` that creates a spring-batch job to reindex selected resources.
`$mark-all-resources-for-reindexing` and `$perform-reindexing-pass` are now deprecated, and will likely be removed in a future release."
```

**Review by michaelabuckley — CHANGES_REQUESTED** [review]
Looks good.  A couple of substantive comments and several nits for you to pick.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/model/RequestPartitionId.java [hunk 7]
```diff
@@ -155,6 +156,14 @@ public int hashCode() {
 			.toHashCode();
 	}
 
+	public String toJson() {
+		try {
+			return ourObjectMapper.writeValueAsString(this);
+		} catch (JsonProcessingException e) {
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/model/RequestPartitionId.java:162]: good idea.  added to JsonUtil

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch/job/PartitionedUrlValidator.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	MatchUrlService myMatchUrlService;
+	@Autowired
+	IRequestPartitionHelperSvc myRequestPartitionHelperSvc;
+	@Autowired
+	FhirContext myFhirContext;
+
+	/**
+	 * This method will throw an exception if the user is not allowed to access the requested resource type on the partition determined by the request
+	 */
+	public List<RequestPartitionId> requestPartitionIdsFromRequestAndUrls(RequestDetails theRequest, List<String> theUrlsToProcess) {
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch/job/PartitionedUrlValidator.java]: Great suggestion.  Code evolved, and you spotted a clean refactoring.  Done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch/job/model/RequestListJson.java [hunk 9]
```diff
@@ -60,8 +60,7 @@ public static RequestListJson fromJson(String theJson) {
 		}
 	}
 
-	@Override
-	public String toString() {
+	public String toJson() {
 		try {
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch/job/model/RequestListJson.java]: done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch/reader/BatchDateThresholdUpdater.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Date;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+
+public class BatchDateThresholdUpdater {
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch/reader/BatchDateThresholdUpdater.java:31]: I considered it but so far the cost is low (reindexing twice) so for now I figure just let it reindex the shutdown batch again.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ProviderConstants.java (hunk 3 above)
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ProviderConstants.java:174]: good point.  For now It's reindex specific.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ReindexProvider.java (hunk 4 above)
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ReindexProvider.java:60]: Good catch!  This one was worth the price of admission!!

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ProviderConstants.java (hunk 2 above)
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ProviderConstants.java:159]: They are different params to different operations.  I'd prefer to keep them separate for now.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ReindexProvider.java (hunk 5 above)
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/ReindexProvider.java:57]: Good catch.  Fixed (when i added the null check)

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch/listener/PidReaderCounterListener.java [hunk 11]
```diff
@@ -40,8 +40,8 @@ public void setStepExecution(StepExecution stepExecution) {
 		myStepExecution = stepExecution;
 	}
 
-	@AfterProcess
-	public void afterProcess(List<Long> thePids, List<String> theSqlList) {
+	@AfterRead
+	public void afterRead(List<Long> thePids) {
 		myTotalPidsProcessed += thePids.size();
 		myStepExecution.getExecutionContext().putLong(RESOURCE_TOTAL_PROCESSED, myTotalPidsProcessed);
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch/listener/PidReaderCounterListener.java:46]: the thing is it doesn't get visible until it's promoted by which time it now means total processed.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch/reader/CronologicalBatchAllResourcePidReader.java [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+		if (myBatchSize == null) {
+			myBatchSize = myDaoConfig.getExpungeBatchSize();
+		}
+		if (executionContext.containsKey(CURRENT_THRESHOLD_LOW)) {
+			myThresholdLow = new Date(executionContext.getLong(CURRENT_THRESHOLD_LOW));
+		}
+	}
+
+	@Override
+	public void update(ExecutionContext executionContext) throws ItemStreamException {
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch/reader/CronologicalBatchAllResourcePidReader.java:126]: Totally agree.  Will fix it when we need to.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IResourceTableDao.java [hunk 13]
```diff
@@ -54,6 +54,9 @@
 	@Query("SELECT t.myId FROM ResourceTable t WHERE t.myUpdated >= :low AND t.myUpdated <= :high ORDER BY t.myUpdated ASC")
 	Slice<Long> findIdsOfResourcesWithinUpdatedRangeOrderedFromOldest(Pageable thePage, @Param("low") Date theLow, @Param("high") Date theHigh);
 
+	@Query("SELECT t.myId FROM ResourceTable t WHERE t.myUpdated >= :low AND t.myUpdated <= :high AND t.myPartitionIdValue = :partition_id ORDER BY t.myUpdated ASC")
+	Slice<Long> findIdsOfPartitionedResourcesWithinUpdatedRangeOrderedFromOldest(Pageable thePage, @Param("low") Date theLow, @Param("high") Date theHigh, @Param("partition_id") Integer theRequestPartitionId);
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IResourceTableDao.java:58]: added TODO

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/DeleteExpungeJobSubmitterImpl.java [hunk 14]
```diff
@@ -63,12 +63,17 @@
...
 
 	@Override
 	@Transactional(Transactional.TxType.NEVER)
-	public JobExecution submitJob(Integer theBatchSize, RequestDetails theRequest, List<String> theUrlsToDeleteExpunge) throws JobParametersInvalidException {
-		List<RequestPartitionId> requestPartitionIds = requestPartitionIdsFromRequestAndUrls(theRequest, theUrlsToDeleteExpunge);
+	public JobExecution submitJob(Integer theBatchSize, List<String> theUrlsToDeleteExpunge, RequestDetails theRequest) throws JobParametersInvalidException {
+		if (theBatchSize == null) {
+			theBatchSize = myDaoConfig.getExpungeBatchSize();
+		}
+		List<RequestPartitionId> requestPartitionIds = myPartitionedUrlValidator.requestPartitionIdsFromRequestAndUrls(theRequest, theUrlsToDeleteExpunge);
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/DeleteExpungeJobSubmitterImpl.java]: done great suggestion

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/job/DeleteExpungeJobConfig.java [hunk 15]
```diff
@@ -56,9 +45,8 @@
  * Delete Expunge job.
  */
 @Configuration
-public class DeleteExpungeJobConfig {
+public class DeleteExpungeJobConfig extends MultiUrlProcessorJobConfig {
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/job/DeleteExpungeJobConfig.java:48]: It's got so little in it, it should arguably just be removed and inlined. There are bean references in the subclasses that wouldn't work across imports.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/reindex/ReindexJobSubmitterImpl.java [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+			throw new ForbiddenOperationException("Reindexing is disabled on this server.");
+		}
+
+		/*
+		 * On the first time we run a particular reindex job, let's make sure we
+		 * have the latest search parameters loaded. A common reason to
+		 * be reindexing is that the search parameters have changed in some way, so
+		 * this makes sure we're on the latest versions
+		 */
+		mySearchParamRegistry.forceRefresh();
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/reindex/ReindexJobSubmitterImpl.java:79]: I wish I could take credit for this, but this comment was literally copy/pasted from the old reindex code

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/reindex/ReindexJobSubmitterImpl.java [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+
+		JobParameters jobParameters = ReverseCronologicalBatchResourcePidReader.buildJobParameters(ProviderConstants.OPERATION_REINDEX, theBatchSize, theUrlsToReindex, requestPartitionIds);
+		return myBatchJobSubmitter.runJob(myReindexJob, jobParameters);
+	}
+
+	@Override
+	@Transactional(Transactional.TxType.NEVER)
+	public JobExecution submitEverythingJob(Integer theBatchSize, RequestDetails theRequest) throws JobParametersInvalidException {
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/reindex/ReindexJobSubmitterImpl.java:87]: I actually did it that way first, but this way ended up being cleaner.  Uncle bob says replace flag parameters with well named methods.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/batch/job/MultiUrlJobParameterUtil.java [hunk 18]
```diff
@@ -18,6 +20,6 @@ public static JobParameters buildJobParameters(String... theUrls) {
 		for (int i = 0; i < theUrls.length; ++i) {
 			requestPartitionIds.add(RequestPartitionId.defaultPartition());
 		}
-		return DeleteExpungeJobConfig.buildJobParameters(2401, Lists.newArrayList(theUrls), requestPartitionIds);
+		return ReverseCronologicalBatchResourcePidReader.buildJobParameters(ProviderConstants.OPERATION_REINDEX, 2401, Lists.newArrayList(theUrls), requestPartitionIds);
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/batch/job/MultiUrlJobParameterUtil.java]: You caught me.  49^2 is my special magic test number.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/batch/reader/BatchDateThresholdUpdaterTest.java [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+	@Test
+	public void threeItemsDifferentELL() {
+		List<Date> dates = Arrays.asList(EARLY_DATE, LATE_DATE, LATE_DATE);
+		mySvc.setDateFromPid(pid -> dates.get(pid.intValue() - 1));
+		Set<Long> seenPids = new HashSet<>();
+		Date newThreshold = mySvc.updateThresholdAndCache(LATE_DATE, seenPids, Arrays.asList(PID1, PID2, PID3));
+		assertEquals(LATE_DATE, newThreshold);
+		assertThat(seenPids, contains(PID2, PID3));
+	}
+
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/batch/reader/BatchDateThresholdUpdaterTest.java:91]: good call.  done

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/storage/IReindexJobSubmitter.java [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+ * #L%
+ */
+
+import ca.uhn.fhir.rest.api.server.RequestDetails;
+import org.springframework.batch.core.JobExecution;
+import org.springframework.batch.core.JobParametersInvalidException;
+
+// Tag interface for Spring wiring
+public interface IReindexJobSubmitter extends IMultiUrlJobSubmitter {
+	JobExecution submitEverythingJob(Integer theBatchSize, RequestDetails theRequest) throws JobParametersInvalidException;
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/api/server/storage/IReindexJobSubmitter.java:29]: it's the submitEverything

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/DeleteExpungeProvider.java (hunk 1 above)
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/DeleteExpungeProvider.java:47]: good suggestion, but this is the last one and I'm eager to get this merge.  I added a TODO

**Pull request merged**